### PR TITLE
Feature notifications

### PR DESCRIPTION
**Pull request by vincent-h-lee**
Attempt #2 

Notifications collected via Ajax on page ready. 
Notifications creation/deletion/updating handled by a NotificationManager. 
Notifications created when Friend request sent, accept, denied (delete notification from request and replace with one for accept or deny). 

Still have the expandable notification class system: different categories and subtypes etc. 
Currently pass a notification as a message as well as the reference_id of the event it is referencing so we can isolate the specific event by cross referencing notification type, category, and reference id. 

Notifications in place for SharedExpense requests -- waiting for that functionality finalization.

### CODE REVIEW

File: app/Http/Controllers/AjaxController.php [hunk 1]
```diff
@@ -46,15 +49,64 @@ public function processFriendRequest(Request $req) {
...
+     * @return an array of arrays. Each sub array contains 1. Notification object, 2. Notification message (string)
+     */
+    public function getNotifications(Request $req)
+    {
+        if($req->ajax()) {
+            $requests = DB::table('notifications')
+                ->where('recipient_username', Auth::user()->username)
+                ->where('category', 2) //category 2 is for just Friend notifications
+                //->where('type', 1) //Just request notifications
+                ->orderBy('is_read', 'ASC')
```
Comment (vincent-h-lee) [inline app/Http/Controllers/AjaxController.php]: sure I'll change that

File: database/migrations/2017_03_25_043016_CreateNotificationsTable.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        Schema::create('notifications', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string('recipient_username');
+            $table->string('sender_username');
+            $table->smallInteger('category');
+            $table->smallInteger('type');
+            $table->string('parameters')->nullable();
+            $table->boolean('is_read')->default(0);
+            $table->integer('reference_id');
+            $table->timestamp('date_added');
```
Comment (vincent-h-lee) [inline database/migrations/2017_03_25_043016_CreateNotificationsTable.php]: Friend request is created >> Notification generated (Category:Friend; Type:Request) 
In this scenario, the reference_id is the id from newly created Friend. The parameter is nullable as not every type of notification needs to pass extra information. For the friend scenario, the parameter could be used when a Friend request is accepted/denied where $notification->parameter ='accepted' or 'denied'.

We have Expense category too and parameters would hold the amount owed in an expense request for example.

File: app/Http/Controllers/AjaxController.php (hunk 1 above)
Comment (nikoootine) [inline app/Http/Controllers/AjaxController.php]: actually this makes sense too. like we want the notifications that are haven't read yet to be on top. just leave it like this for now and then let's judge after

**Comment by itsmattli** [comment]
looks good, but lets keep track of what is still needed in the notification feature

**Pull request merged**